### ISSUE TITLE
calibrate_qwen

### ISSUE THREAD

**Issue by ritwikraha** [ritwikraha/AutoRegressive-Bhasha#1]
## Distilling **uncertainty-aware multiple-choice reasoning**

Train a small Qwen model to answer difficult multiple-choice questions while producing:

1. the selected answer;
2. a short justification;
3. a calibrated confidence score;
4. an abstention when confidence is genuinely low.

Example output:

```json
{
  "answer": "C",
  "confidence": 0.72,
  "justification": "C is correct because the intervention blocks the causal pathway described in the question.",
  "abstain": false
}
```

The central research question would be:
> Does on-policy distillation transfer not only answer accuracy, but also useful uncertainty estimates and selective prediction behaviour from a stronger teacher to a smaller student?

This is lightweight enough for a first Tinker project, but substantial enough to become a useful open-source report, dataset and benchmark.

---

# Why this is better than ordinary reasoning distillation

A standard "distil Qwen on GSM8K" project is now too common. It would teach you Tinker, but it would feel like a tutorial reproduction.

Adding **confidence calibration and abstention** creates a meaningful ML problem:

- Small models are often confidently wrong.
- Teacher probabilities do not automatically transfer through ordinary SFT.
- Verbal confidence such as "90%" is frequently poorly calibrated.
- On-policy examples may help specifically on inputs where the student is uncertain or incorrect.
- The final model is useful in routing, ensembles, retrieval systems and human-review pipelines.

The outcome is not just "a smaller model with higher benchmark accuracy." It becomes:
> A small model that knows more accurately when it is likely to be wrong.

That is broadly useful across ML applications.

---

## **Know When You Don't Know: On-Policy Distillation of Calibrated Reasoning into Small Language Models**

A more repository-friendly name:

```
CalibrateQwen
```

---

# What the project teaches you about Tinker

This project can cover almost the entire useful Tinker workflow:

| Stage                    | What you learn                                             |
| ------------------------ | ---------------------------------------------------------- |
| Sampling                 | Generate student and teacher responses                     |
| SFT                      | Train on teacher-generated solutions                       |
| Soft-target distillation | Use teacher token probabilities                            |
| On-policy distillation   | Train the teacher on student-generated trajectories        |
| Custom losses            | Weight answer, rationale and confidence tokens differently |
| Evaluation               | Run periodic benchmark evaluation                          |
| Checkpointing            | Save and compare intermediate models                       |
| Model export             | Download the LoRA adapter or merged model                  |
| Ablations                | Compare multiple training recipes                          |

[Tinker provides official recipes](https://tinker-docs.thinkingmachines.ai/tinker/lora-primer/) for both off-policy and on-policy model distillation. Its current training approach is LoRA-based, which is well suited to relatively small post-training datasets.

---

# Scope

## Student

Use a compact Qwen base or instruct model in the **1.5B–4B range**, selected from Tinker's current supported-model list at the time you run the experiment.

A 1.5B model gives a more visible distillation gap. A 4B model gives stronger final results.

My preferred setup:

```
Student: compact Qwen 1.5B–4B
Teacher: Qwen model around 9B–27B
Training: LoRA through Tinker
```

Do not hard-code an older model identifier from a blog post. Tinker's model lineup changes, and several earlier Qwen variants were retired in June and July 2026. Check the [live model catalogue](https://tinker-docs.thinkingmachines.ai/tinker/models/?utm_source=chatgpt.com) before launching.

## Dataset size

Start with:

```
Training questions: 8,000–15,000
Validation questions: 1,000
Held-out test questions: 2,000–4,000
```

That is large enough to support real comparisons while remaining manageable.

## Sequence length

Keep responses below approximately:

```
128–300 output tokens
```

This matters because token generation and teacher scoring will drive much of the cost.

---

# Dataset design

Use a mixture of public multiple-choice datasets rather than one benchmark.

Possible categories:

- science reasoning;
- commonsense reasoning;
- factual knowledge;
- causal reasoning;
- logical reasoning;
- basic quantitative reasoning.

Potential datasets include:

- ARC Challenge;
- OpenBookQA;
- CommonsenseQA;
- MMLU subsets;
- StrategyQA converted to constrained choices;
- synthetic counterfactual questions;
- adversarially perturbed variants.

Avoid training and evaluating on random splits from exactly the same source alone. Otherwise, you may mostly measure dataset-style adaptation.

A better split is:

```
Train:
ARC + OpenBookQA + selected MMLU subjects

In-domain test:
Held-out examples from those datasets

Out-of-domain test:
CommonsenseQA + unseen MMLU subjects

Perturbation test:
Paraphrased and option-shuffled questions
```

---

# The key experimental comparison

Run four student variants.

## 1. Base model

No training.

This establishes:

- initial accuracy;
- calibration;
- abstention quality;
- formatting reliability.

## 2. Hard-label SFT

Train only on:

```json
{
  "answer": "C",
  "confidence": 1.0
}
```

This is the simplest baseline.

It will probably improve accuracy but may produce overconfidence.

## 3. Off-policy teacher distillation

The teacher generates answers and rationales before student training.

The student learns from fixed teacher outputs:

```
Dataset prompt
        ↓
Teacher completion
        ↓
Fixed training example
        ↓
Student SFT/distillation
```

This is easy and stable, but the teacher never sees the mistakes the student naturally generates.

## 4. On-policy distillation

The student first samples an answer. The teacher then scores or corrects that student trajectory.

```
Question
   ↓
Student-generated response
   ↓
Teacher token distribution or critique
   ↓
Student update
```

This directly targets the regions of output space occupied by the student.

Thinking Machines' published [on-policy distillation work](https://thinkingmachines.ai/blog/on-policy-distillation/?utm_source=chatgpt.com) used the same broad principle for mathematical reasoning: sample from the student, then use a stronger teacher to supervise those student-generated sequences. Their original Qwen3-8B and Qwen3-32B models have since been retired from Tinker, but the algorithmic recipe remains available.

---

# Make uncertainty supervision meaningful

Do not simply ask the teacher to invent a confidence percentage. That produces noisy labels.

Instead, estimate teacher confidence using repeated sampling or probability aggregation.

## Method A: Teacher answer probability

For each option $a_i$, obtain a normalized teacher score:

$$
p_T(a_i \mid x) = \frac{\exp(s_i/\tau)}{\sum_j \exp(s_j/\tau)}
$$

Where:

- $s_i$ is the teacher score for option $i$;
- $\tau$ is a temperature;
- $x$ is the question.

Train the student's confidence to approximate:

$$
c_T = p_T(a^* \mid x)
$$

where $a^*$ is the teacher's selected answer.

## Method B: Teacher sampling agreement

Sample the teacher $K$ times:

$$
c_T = \frac{1}{K} \sum_{k=1}^{K} \mathbf{1}[a_k = a_{\text{mode}}]
$$

For example, with five samples:

```
C, C, C, B, C
```

Teacher agreement is:

```
4 / 5 = 0.80
```

This is more expensive but simple and interpretable.

## Method C: Correctness-conditioned confidence

Where gold labels are available, combine teacher agreement and correctness:

```
Teacher correct + high agreement → high confidence
Teacher wrong + high agreement   → dangerous overconfidence example
Low agreement                    → uncertain example
```

Those confidently wrong examples are especially valuable for studying whether teacher uncertainty is actually worth transferring.

---

# Abstention policy

Define a threshold $\gamma$:

$$
\text{abstain}(x) = \mathbf{1}[\hat{c}(x) < \gamma]
$$

Evaluate several thresholds:

```
0.50
0.60
0.70
0.80
0.90
```

This produces a risk–coverage curve.

- **Coverage**: fraction of questions answered.
- **Risk**: error rate among answered questions.

A useful distilled model should reduce error substantially as it abstains on more examples.

---

# Evaluation metrics

## Accuracy

Standard multiple-choice accuracy.

## Expected Calibration Error

Partition predictions into confidence bins:

$$
\operatorname{ECE} = \sum_{m=1}^{M} \frac{|B_m|}{n} \left| \operatorname{acc}(B_m) - \operatorname{conf}(B_m) \right|
$$

Lower is better.

## Brier score

For predicted option probabilities:

$$
\operatorname{BS} = \frac{1}{N} \sum_{i=1}^{N} \sum_{k=1}^{K} (p_{ik} - y_{ik})^2
$$

Lower is better.

## Negative log-likelihood

Measures whether probability assigned to the correct answer is high.

## Selective accuracy

Accuracy after abstaining below a confidence threshold.

## Area under risk–coverage curve

Measures confidence ranking across all possible thresholds.

## Overconfidence on errors

Report:

$$
\mathbb{E}[\hat{c} \mid \hat{y} \neq y]
$$

This may be the most intuitively important result.

## Format validity

Percentage of outputs that satisfy the JSON schema.

---

# A particularly valuable ablation

Compare three confidence representations.

### Numeric verbal confidence

```json
"confidence": 0.73
```

### Discrete confidence buckets

```json
"confidence": "medium"
```

Where:

```
low    = 0.00–0.50
medium = 0.50–0.80
high   = 0.80–1.00
```

### Answer-token probabilities

Do not ask the model to state confidence. Derive confidence directly from the probability assigned to answer tokens.

This can answer a useful research question:
> Is explicitly verbalized confidence more or less calibrated than confidence extracted from model probabilities after distillation?

That is publishable-quality analysis if performed carefully.

---

# Lightweight training plan

## Phase 1: Learn Tinker with SFT

Use 2,000 examples.

Train the model to return valid structured outputs:

```json
{
  "answer": "A",
  "confidence": 0.64,
  "justification": "...",
  "abstain": false
}
```

Goal:

- verify tokenization;
- understand renderers;
- create batches;
- train LoRA;
- sample checkpoints;
- export weights.

Tinker has [official introductory workflows for sampling, SFT, rendering and exporting](https://tinker-docs.thinkingmachines.ai/tutorials/basics/hello-tinker/?utm_source=chatgpt.com) LoRA or merged Hugging Face models.

## Phase 2: Off-policy distillation

Increase to 8,000–10,000 examples.

Use fixed teacher outputs or soft teacher targets.

Goal:

- establish an accuracy and calibration baseline;
- understand teacher/student configuration;
- validate the official distillation recipe.

## Phase 3: On-policy distillation

Use a smaller set of approximately 3,000–6,000 questions.

For every training batch:

1. sample one or two responses from the student;
2. evaluate those sequences with the teacher;
3. calculate the distillation objective;
4. update the student.

Goal:

- learn Tinker's sampling-plus-training loop;
- test whether on-policy supervision improves student-specific errors.

## Phase 4: Calibration pass

Use a held-out calibration set to choose either:

- temperature scaling;
- an abstention threshold;
- confidence buckets;
- a small calibration head outside the LLM.

Do not tune this on the final test set.

---

# Keep the project economical

The biggest cost controls are:

- use short rationales;
- limit each question to four answer options;
- sample one student trajectory initially;
- use teacher rescoring instead of multiple long teacher generations;
- run evaluation on subsets during training;
- use 2,000 examples for pipeline debugging;
- train the full experiment only after all parsing and metrics work;
- avoid reinforcement learning in version one.

The initial project does **not** need:

- a reward model;
- GRPO;
- long chain-of-thought;
- tool execution;
- multimodal input;
- 100,000 examples;
- full model fine-tuning.

[Tinker handles distributed training remotely and exposes SFT and RL workflows through its SDK](https://tinker-docs.thinkingmachines.ai/tinker/?utm_source=chatgpt.com), but for this project, supervised and distillation recipes are sufficient.

---

# Estimated experiment matrix

A credible first release could contain:

| Run | Method                       | Examples | Purpose               |
| --- | ---------------------------- | -------- | --------------------- |
| A   | Base model                   | 0        | Baseline              |
| B   | Hard-label SFT               | 8k       | Accuracy baseline     |
| C   | Teacher completion SFT       | 8k       | Standard distillation |
| D   | Off-policy soft distillation | 8k       | Probability transfer  |
| E   | On-policy distillation       | 4k–8k    | Main method           |
| F   | SFT → on-policy              | 8k + 4k  | Combined recipe       |

Then run confidence-format ablations only on the two strongest methods.

This is enough for a serious technical report without exploding the total number of runs.

---

# Expected deliverables

Release:

```
calibrate-qwen/
├── data/
│   ├── build_dataset.py
│   ├── perturb_options.py
│   └── generate_teacher_targets.py
├── training/
│   ├── train_sft.py
│   ├── train_off_policy.py
│   └── train_on_policy.py
├── evaluation/
│   ├── accuracy.py
│   ├── calibration.py
│   ├── selective_prediction.py
│   └── robustness.py
├── configs/
├── notebooks/
├── results/
└── README.md
```

Publish:

- the processed dataset;
- teacher-confidence metadata;
- the Tinker training configuration;
- student LoRA adapter;
- evaluation scripts;
- risk–coverage plots;
- full ablation table;
- cost and token-use report.

[Tinker-trained adapters can be downloaded, converted to PEFT format and published to Hugging Face](https://tinker-docs.thinkingmachines.ai/tutorials/deployment/publish-hub/?utm_source=chatgpt.com), which makes the resulting artefact reusable outside Tinker.

---

# What would make the project high-signal

The strongest result is not necessarily a large increase in accuracy.

A compelling result could look like:

```
Method                    Accuracy    ECE    Error confidence    80% coverage accuracy
Base                        54.2%     .19          .71                   63.4%
Teacher-output SFT          61.8%     .17          .75                   70.1%
Off-policy distillation     63.0%     .12          .63                   74.8%
On-policy distillation      64.1%     .07          .49                   80.3%
```

Even a moderate accuracy gain would be meaningful if on-policy distillation substantially improves:

- calibration;
- error detection;
- abstention;
- out-of-domain reliability.

## Final recommendation

Build **CalibrateQwen** using a compact Qwen student, 8,000–15,000 mixed-domain multiple-choice questions, and a stronger Qwen teacher. Compare hard-label SFT, off-policy distillation and on-policy distillation, with calibration and selective prediction as the central contribution.

It is light enough to complete, broad enough to interest the ML community, and exercises the part of Tinker that is most scientifically distinctive: training a student from teacher distributions on the student's own outputs.
